File: scripts/export-mcp-so.js
#!/usr/bin/env node

/**
 * Browser Console Script to Export mcp.so Data
 * 
 * Instructions:
 * 1. Open https://mcp.so/servers?tag=latest in your browser
 * 2. Open Developer Console (F12)
 * 3. Copy and paste this entire script into the console
 * 4. Wait for it to complete (it will auto-scroll through pages)
 * 5. Save the output JSON file
 */

// Configuration
const DELAY_BETWEEN_PAGES = 2000; // 2 seconds
const MAX_PAGES = 510; // Adjust based on actual page count

// Data storage
const allServers = [];

// Extract servers from current page
function extractServersFromPage() {
    const servers = [];
    
    // Method 1: Extract from server links
    document.querySelectorAll('a[href*="/server/"]').forEach(link => {
        const href = link.getAttribute('href');
        if (!href || href === '/servers') return;
        
        // Get parent container
        const container = link.closest('div[class*="card"], article, .server-item, [class*="grid"] > div');
        
        // Extract server info
        const server = {
            name: link.textContent.trim(),
            url: href.startsWith('http') ? href : `https://mcp.so${href}`,
            extractedAt: new Date().toISOString()
        };
        
        // Try to get description
        if (container) {
            const descriptions = container.querySelectorAll('p, div[class*="description"], div[class*="text-muted"]');
            if (descriptions.length > 0) {
                server.description = descriptions[0].textContent.trim();
            }
            
            // Try to get author
            const authorElements = container.querySelectorAll('span[class*="author"], div[class*="author"], a[href*="/user/"]');
            if (authorElements.length > 0) {
                server.author = authorElements[0].textContent.trim();
            }
            
            // Try to get tags
            const tagElements = container.querySelectorAll('span[class*="tag"], span[class*="badge"], div[class*="tag"]');
            if (tagElements.length > 0) {
                server.tags = Array.from(tagElements).map(el => el.textContent.trim());
            }
            
            // Look for install commands in the container
            const codeBlocks = container.querySelectorAll('code, pre');
            codeBlocks.forEach(code => {
                const text = code.textContent;
                if (text.includes('npx') || text.includes('npm install') || text.includes('uvx')) {
                    server.installCommand = text.trim();
                }
            });
        }
        
        // Parse URL to get slug and author
        const urlParts = href.split('/');
        if (urlParts.length >= 2) {
            server.slug = urlParts[urlParts.length - 2];
            server.authorFromUrl = urlParts[urlParts.length - 1];
        }
        
        servers.push(server);
    });
    
    // Method 2: Try to extract from React/Next.js data
    try {
        // Look for __NEXT_DATA__ script tag
        const scripts = document.querySelectorAll('script#__NEXT_DATA__');
        scripts.forEach(script => {
            try {
                const data = JSON.parse(script.textContent);
                if (data.props?.pageProps?.servers) {
                    console.log('Found servers in __NEXT_DATA__');
                    servers.push(...data.props.pageProps.servers);
                }
            } catch (e) {
                // Ignore parse errors
            }
        });
    } catch (e) {
        console.log('Could not extract from __NEXT_DATA__');
    }
    
    // Deduplicate
    const uniqueServers = [];
    const seen = new Set();
    
    servers.forEach(server => {
        const key = server.url || server.name;
        if (!seen.has(key)) {
            seen.add(key);
            uniqueServers.push(server);
        }
    });
    
    return uniqueServers;
}

// Navigate to next page
async function goToNextPage(currentPage) {
    const nextPage = currentPage + 1;
    const nextUrl = `https://mcp.so/servers?tag=latest&page=${nextPage}`;
    
    // Try clicking next button first
    const nextButtons = document.querySelectorAll('a[href*="page=' + nextPage + '"], button:contains("Next"), a:contains("Next")');
    if (nextButtons.length > 0) {
        nextButtons[0].click();
    } else {
        // Fallback to direct navigation
        window.location.href = nextUrl;
    }
    
    // Wait for page to load
    return new Promise(resolve => {
        setTimeout(resolve, DELAY_BETWEEN_PAGES);
    });
}

// Main extraction function
async function extractAllServers() {
    console.log('Starting mcp.so data extraction...');
    console.log('This will take several minutes. Please keep this tab active.');
    
    let currentPage = 1;
    let foundServers = true;
    
    while (foundServers && currentPage <= MAX_PAGES) {
        console.log(`Extracting page ${currentPage}...`);
        
        // Extract from current page
        const pageServers = extractServersFromPage();
        console.log(`Found ${pageServers.length} servers on page ${currentPage}`);
        
        if (pageServers.length === 0) {
            // Try one more time after a delay
            await new Promise(resolve => setTimeout(resolve, 1000));
            const retryServers = extractServersFromPage();
            if (retryServers.length === 0) {
                console.log('No more servers found. Extraction complete.');
                foundServers = false;
                break;
            } else {
                allServers.push(...retryServers);
            }
        } else {
            allServers.push(...pageServers);
        }
        
        // Go to next page
        if (foundServers && currentPage < MAX_PAGES) {
            await goToNextPage(currentPage);
            currentPage++;
        } else {
            break;
        }
    }
    
    // Final deduplication
    const finalServers = [];
    const seen = new Set();
    
    allServers.forEach(server => {
        const key = server.url || server.name;
        if (!seen.has(key)) {
            seen.add(key);
            finalServers.push(server);
        }
    });
    
    console.log(`\nExtraction complete!`);
    console.log(`Total servers found: ${finalServers.length}`);
    console.log(`\nTo save the data:`);
    console.log(`1. Copy the JSON output below`);
    console.log(`2. Save it as 'mcp-so-servers.json'`);
    console.log(`3. Run: node src/scrapers/mcp-so-bulk-analyzer.js analyze mcp-so-servers.json`);
    
    // Output the data
    console.log('\n=== JSON DATA START ===');
    console.log(JSON.stringify(finalServers, null, 2));
    console.log('=== JSON DATA END ===');
    
    // Also try to trigger download
    try {
        const blob = new Blob([JSON.stringify(finalServers, null, 2)], { type: 'application/json' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'mcp-so-servers.json';
        a.click();
        console.log('\nFile download triggered!');
    } catch (e) {
        console.log('\nCould not trigger automatic download. Please copy the JSON manually.');
    }
    
    return finalServers;
}

// Manual single-page extraction (if auto-navigation doesn't work)
function extractCurrentPageOnly() {
    const servers = extractServersFromPage();
    console.log(`Found ${servers.length} servers on current page`);
    console.log(JSON.stringify(servers, null, 2));
    return servers;
}

// Start extraction
console.log('==============================================');
console.log('mcp.so Data Extractor');
console.log('==============================================');
console.log('Commands:');
console.log('- extractAllServers()   : Extract from all pages (auto-navigate)');
console.log('- extractCurrentPageOnly() : Extract from current page only');
console.log('==============================================');
console.log('\nTo start full extraction, run: extractAllServers()');

File: scripts/export-mcp-so.py
#!/usr/bin/env python3
"""
Export mcp.so data using Selenium with undetected-chromedriver
This can often bypass Cloudflare protection

Requirements:
pip install selenium undetected-chromedriver beautifulsoup4
"""

import json
import time
import sys
from typing import List, Dict
import undetected_chromedriver as uc
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from bs4 import BeautifulSoup

class McpSoExporter:
    def __init__(self):
        self.base_url = "https://mcp.so/servers"
        self.servers = []
        self.driver = None
        
    def setup_driver(self):
        """Setup undetected Chrome driver"""
        options = uc.ChromeOptions()
        # options.add_argument('--headless')  # Uncomment for headless mode
        options.add_argument('--disable-blink-features=AutomationControlled')
        
        self.driver = uc.Chrome(options=options)
        self.driver.implicitly_wait(10)
        
    def wait_for_cloudflare(self):
        """Wait for Cloudflare challenge to complete"""
        time.sleep(5)  # Initial wait
        
        # Check if we're still on a Cloudflare challenge page
        try:
            if "just a moment" in self.driver.page_source.lower():
                print("Waiting for Cloudflare challenge...")
                time.sleep(10)  # Wait longer
        except:
            pass
            
    def extract_servers_from_page(self) -> List[Dict]:
        """Extract server information from current page"""
        servers = []
        soup = BeautifulSoup(self.driver.page_source, 'html.parser')
        
        # Try multiple selectors
        server_links = soup.find_all('a', href=lambda x: x and '/server/' in x)
        
        for link in server_links:
            href = link.get('href', '')
            if href == '/servers':
                continue
                
            # Get parent container
            container = link.find_parent('div', class_=lambda x: x and any(
                cls in str(x) for cls in ['card', 'item', 'grid']
            ))
            
            server = {
                'name': link.get_text(strip=True),
                'url': f"https://mcp.so{href}" if not href.startswith('http') else href,
            }
            
            if container:
                # Extract description
                desc_elem = container.find(['p', 'div'], class_=lambda x: x and 'desc' in str(x).lower())
                if desc_elem:
                    server['description'] = desc_elem.get_text(strip=True)
                
                # Extract author
                author_elem = container.find(['span', 'div'], class_=lambda x: x and 'author' in str(x).lower())
                if author_elem:
                    server['author'] = author_elem.get_text(strip=True)
                
                # Look for install commands
                code_blocks = container.find_all(['code', 'pre'])
                for code in code_blocks:
                    text = code.get_text(strip=True)
                    if any(cmd in text for cmd in ['npx', 'npm install', 'uvx']):
                        server['installCommand'] = text
                        break
            
            # Parse URL for additional info
            url_parts = href.strip('/').split('/')
            if len(url_parts) >= 3:
                server['slug'] = url_parts[-2]
                server['authorFromUrl'] = url_parts[-1]
            
            servers.append(server)
        
        # Try to extract from JavaScript data
        scripts = soup.find_all('script', id='__NEXT_DATA__')
        for script in scripts:
            try:
                data = json.loads(script.string)
                if 'props' in data and 'pageProps' in data['props']:
                    if 'servers' in data['props']['pageProps']:
                        print("Found servers in __NEXT_DATA__")
                        servers.extend(data['props']['pageProps']['servers'])
            except:
                pass
        
        # Deduplicate
        seen = set()
        unique_servers = []
        for server in servers:
            key = server.get('url', server.get('name', ''))
            if key and key not in seen:
                seen.add(key)
                unique_servers.append(server)
        
        return unique_servers
    
    def navigate_all_pages(self, max_pages: int = 510):
        """Navigate through all pages and extract data"""
        current_page = 1
        
        # Go to first page
        url = f"{self.base_url}?tag=latest"
        print(f"Loading {url}")
        self.driver.get(url)
        self.wait_for_cloudflare()
        
        while current_page <= max_pages:
            print(f"\nProcessing page {current_page}...")
            
            # Wait for content to load
            try:
                WebDriverWait(self.driver, 10).until(
                    EC.presence_of_element_located((By.TAG_NAME, "a"))
                )
            except:
                print("Timeout waiting for page to load")
            
            # Extract servers
            page_servers = self.extract_servers_from_page()
            print(f"Found {len(page_servers)} servers on page {current_page}")
            
            if not page_servers:
                print("No servers found, checking if we've reached the end...")
                time.sleep(2)
                page_servers = self.extract_servers_from_page()
                if not page_servers:
                    print("No more servers, stopping.")
                    break
            
            self.servers.extend(page_servers)
            
            # Try to go to next page
            next_url = f"{self.base_url}?tag=latest&page={current_page + 1}"
            
            try:
                # Method 1: Click next button
                next_button = self.driver.find_element(
                    By.XPATH, 
                    f"//a[contains(@href, 'page={current_page + 1}')]"
                )
                next_button.click()
                time.sleep(2)
            except:
                # Method 2: Direct navigation
                self.driver.get(next_url)
                time.sleep(2)
            
            current_page += 1
        
        print(f"\nExtraction complete. Total servers: {len(self.servers)}")
    
    def save_data(self, filename: str = "mcp-so-servers.json"):
        """Save extracted data to JSON file"""
        # Deduplicate final list
        seen = set()
        unique_servers = []
        
        for server in self.servers:
            key = server.get('url', server.get('name', ''))
            if key and key not in seen:
                seen.add(key)
                unique_servers.append(server)
        
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(unique_servers, f, indent=2, ensure_ascii=False)
        
        print(f"\nSaved {len(unique_servers)} unique servers to {filename}")
        
    def extract_single_page(self, page_num: int = 1):
        """Extract data from a single page"""
        url = f"{self.base_url}?tag=latest" if page_num == 1 else f"{self.base_url}?tag=latest&page={page_num}"
        print(f"Loading {url}")
        
        self.driver.get(url)
        self.wait_for_cloudflare()
        
        servers = self.extract_servers_from_page()
        print(f"Found {len(servers)} servers")
        return servers
    
    def run(self, mode: str = "all", max_pages: int = 510):
        """Run the exporter"""
        try:
            self.setup_driver()
            
            if mode == "all":
                self.navigate_all_pages(max_pages)
                self.save_data()
            elif mode == "single":
                servers = self.extract_single_page()
                print(json.dumps(servers, indent=2))
            
        finally:
            if self.driver:
                self.driver.quit()

def main():
    import argparse
    
    parser = argparse.ArgumentParser(description="Export mcp.so server data")
    parser.add_argument(
        "--mode", 
        choices=["all", "single"], 
        default="all",
        help="Extract all pages or just current page"
    )
    parser.add_argument(
        "--max-pages",
        type=int,
        default=510,
        help="Maximum number of pages to extract"
    )
    parser.add_argument(
        "--output",
        default="mcp-so-servers.json",
        help="Output filename"
    )
    
    args = parser.parse_args()
    
    exporter = McpSoExporter()
    
    print("==============================================")
    print("mcp.so Data Exporter (Selenium)")
    print("==============================================")
    print(f"Mode: {args.mode}")
    print(f"Max pages: {args.max_pages}")
    print(f"Output: {args.output}")
    print("==============================================\n")
    
    try:
        exporter.run(mode=args.mode, max_pages=args.max_pages)
        
        if args.mode == "all" and args.output != "mcp-so-servers.json":
            import shutil
            shutil.move("mcp-so-servers.json", args.output)
            
        print("\nNext steps:")
        print(f"1. Check the exported file: {args.output}")
        print(f"2. Analyze the data: node src/scrapers/mcp-so-bulk-analyzer.js analyze {args.output}")
        
    except KeyboardInterrupt:
        print("\nExport interrupted by user")
        if exporter.servers:
            exporter.save_data("mcp-so-partial.json")
            print("Partial data saved to mcp-so-partial.json")
    except Exception as e:
        print(f"\nError: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()